File: commonutils/src/main/java/net/maxsmr/commonutils/logger/BaseLogger.java
package net.maxsmr.commonutils.logger;

import android.support.annotation.NonNull;

import net.maxsmr.commonutils.data.Predicate;

import java.util.Collection;

public abstract class BaseLogger {

    private boolean isLoggingEnabled = true;

    public boolean isLoggingEnabled() {
        return isLoggingEnabled;
    }

    public void setLoggingEnabled(boolean toggle) {
        this.isLoggingEnabled = toggle;
    }

    public abstract void v(String message);

    public abstract void v(Throwable exception);

    public abstract void v(String message, Throwable exception);

    public abstract void d(String message);

    public abstract void d(Throwable exception);

    public abstract void d(String message, Throwable exception);

    public abstract void i(String message);

    public abstract void i(Throwable exception);

    public abstract void i(String message, Throwable exception);

    public abstract void w(String message);

    public abstract void w(Throwable exception);

    public abstract void w(String message, Throwable exception);

    public abstract void e(String message);

    public abstract void e(Throwable exception);

    public abstract void e(String message, Throwable exception);

    public abstract void wtf(String message);

    public abstract void wtf(Throwable exception);

    public abstract void wtf(String message, Throwable exception);

    public enum Level {

        VERBOSE, DEBUG, INFO, WARN, ERROR, WTF;

        public boolean isVerbose() {
            return isInRange(VERBOSE);
        }

        public boolean isDebug() {
            return isInRange(DEBUG);
        }

        public boolean isInfo() {
            return isInRange(INFO);
        }

        public boolean isWarn() {
            return isInRange(WARN);
        }

        public boolean isError() {
            return isInRange(ERROR);
        }

        public boolean isWtf() {
            return isInRange(WTF);
        }

        private boolean isInRange(@NonNull Level level) {
            return ordinal() <= level.ordinal();
        }

        public static boolean contains(final Level level, Collection<Level> levels) {
            return Predicate.Methods.contains(levels, element -> element == level);
        }
    }

    public static class Stub extends BaseLogger {

        @Override
        public void v(String message) {

        }

        @Override
        public void v(Throwable exception) {

        }

        @Override
        public void v(String message, Throwable exception) {

        }

        @Override
        public void d(String message) {

        }

        @Override
        public void d(Throwable exception) {

        }

        @Override
        public void d(String message, Throwable exception) {

        }

        @Override
        public void i(String message) {

        }

        @Override
        public void i(Throwable exception) {

        }

        @Override
        public void i(String message, Throwable exception) {

        }

        @Override
        public void w(String message) {

        }

        @Override
        public void w(Throwable exception) {

        }

        @Override
        public void w(String message, Throwable exception) {

        }

        @Override
        public void e(String message) {

        }

        @Override
        public void e(Throwable exception) {

        }

        @Override
        public void e(String message, Throwable exception) {

        }

        @Override
        public void wtf(String message) {

        }

        @Override
        public void wtf(Throwable exception) {

        }

        @Override
        public void wtf(String message, Throwable exception) {

        }
    }
}

File: commonutils-jre/src/main/kotlin/net/maxsmr/commonutils/logger/BaseLogger.kt
package net.maxsmr.commonutils.logger

import net.maxsmr.commonutils.logger.holder.BaseLoggerHolder

abstract class BaseLogger(protected val tag: String) {

    abstract fun _v(message: String)

    abstract fun _v(exception: Throwable)

    abstract fun _v(message: String, exception: Throwable?)

    abstract fun _d(message: String)

    abstract fun _d(exception: Throwable)

    abstract fun _d(message: String, exception: Throwable?)

    abstract fun _i(message: String)

    abstract fun _i(exception: Throwable)

    abstract fun _i(message: String, exception: Throwable?)

    abstract fun _w(message: String)

    abstract fun _w(exception: Throwable)

    abstract fun _w(message: String, exception: Throwable?)

    abstract fun _e(message: String)

    abstract fun _e(exception: Throwable)

    abstract fun _e(message: String, exception: Throwable?)

    abstract fun _wtf(message: String)

    abstract fun _wtf(exception: Throwable)

    abstract fun _wtf(message: String, exception: Throwable?)

    fun v(message: String) {
        if (BaseLoggerHolder.instance.loggerLevel.isVerbose) {
            _v(message)
        }
    }

    fun v(exception: Throwable) {
        if (BaseLoggerHolder.instance.loggerLevel.isVerbose) {
            _v(exception)
        }
    }

    fun v(message: String, exception: Throwable?) {
        if (BaseLoggerHolder.instance.loggerLevel.isVerbose) {
            _v(message, exception)
        }
    }

    fun d(message: String) {
        if (BaseLoggerHolder.instance.loggerLevel.isDebug) {
            _d(message)
        }
    }

    fun d(exception: Throwable) {
        if (BaseLoggerHolder.instance.loggerLevel.isDebug) {
            _d(exception)
        }
    }

    fun d(message: String, exception: Throwable?) {
        if (BaseLoggerHolder.instance.loggerLevel.isDebug) {
            _d(message, exception)
        }
    }

    fun i(message: String) {
        if (BaseLoggerHolder.instance.loggerLevel.isInfo) {
            _i(message)
        }
    }

    fun i(exception: Throwable) {
        if (BaseLoggerHolder.instance.loggerLevel.isInfo) {
            _i(exception)
        }
    }

    fun i(message: String, exception: Throwable?) {
        if (BaseLoggerHolder.instance.loggerLevel.isInfo) {
            _i(message, exception)
        }
    }

    fun w(message: String) {
        if (BaseLoggerHolder.instance.loggerLevel.isWarn) {
            _w(message)
        }
    }

    fun w(exception: Throwable) {
        if (BaseLoggerHolder.instance.loggerLevel.isWarn) {
            _w(exception)
        }
    }

    fun w(message: String, exception: Throwable?) {
        if (BaseLoggerHolder.instance.loggerLevel.isWarn) {
            _w(message, exception)
        }
    }

    fun e(message: String) {
        if (BaseLoggerHolder.instance.loggerLevel.isError) {
            _e(message)
        }
    }

    fun e(exception: Throwable) {
        if (BaseLoggerHolder.instance.loggerLevel.isError) {
            _e(exception)
        }
    }

    fun e(message: String, exception: Throwable?) {
        if (BaseLoggerHolder.instance.loggerLevel.isError) {
            _e(message, exception)
        }
    }

    fun wtf(message: String) {
        if (BaseLoggerHolder.instance.loggerLevel.isWtf) {
            _wtf(message)
        }
    }

    fun wtf(exception: Throwable) {
        if (BaseLoggerHolder.instance.loggerLevel.isWtf) {
            _wtf(exception)
        }
    }

    fun wtf(message: String, exception: Throwable?) {
        if (BaseLoggerHolder.instance.loggerLevel.isWtf) {
            _wtf(message, exception)
        }
    }

    enum class Level {
        VERBOSE, DEBUG, INFO, WARN, ERROR, WTF;

        val isVerbose: Boolean
            get() = isInRange(VERBOSE)

        val isDebug: Boolean
            get() = isInRange(DEBUG)

        val isInfo: Boolean
            get() = isInRange(INFO)

        val isWarn: Boolean
            get() = isInRange(WARN)

        val isError: Boolean
            get() = isInRange(ERROR)

        val isWtf: Boolean
            get() = isInRange(WTF)

        private fun isInRange(level: Level): Boolean {
            return ordinal <= level.ordinal
        }
    }

    class Stub(tag: String) : BaseLogger(tag) {

        override fun _v(message: String) {
        }

        override fun _v(exception: Throwable) {
        }

        override fun _v(message: String, exception: Throwable?) {
        }

        override fun _d(message: String) {
        }

        override fun _d(exception: Throwable) {
        }

        override fun _d(message: String, exception: Throwable?) {
        }

        override fun _i(message: String) {
        }

        override fun _i(exception: Throwable) {
        }

        override fun _i(message: String, exception: Throwable?) {
        }

        override fun _w(message: String) {
        }

        override fun _w(exception: Throwable) {
        }

        override fun _w(message: String, exception: Throwable?) {
        }

        override fun _e(message: String) {
        }

        override fun _e(exception: Throwable) {
        }

        override fun _e(message: String, exception: Throwable?) {
        }

        override fun _wtf(message: String) {
        }

        override fun _wtf(exception: Throwable) {
        }

        override fun _wtf(message: String, exception: Throwable?) {
        }
    }
}
